✨ zv: Added a signature! macro

hadihaider055 opened the pull request:
Add a `signature!` macro that creates const signatures with compile-time validation. This makes it easier to work with D-Bus signatures while catching invalid ones early.

Fixes #984

zeenix commented:
> @zeenix, I have moved the implementation to the `zvariant_utils` and re-exported it from the `zvariant`. Fixed all the issues related to `pub(crate)`, and the impl is split and moved to a separate [file](https://github.com/z-galaxy/zbus/pull/1536/files#diff-17bc425b1898f55730e4427419eab0521d5ec9ce723f9353ff2acbb3bafc7376) in `zvariant_utils`.

Thanks for doing that. :+1: However, the macro still remains in zvariant_derive. The idea was that it doesn't belong there, so it should only be zvariant that rexports it and the whole macro (not just the impl) should be in zvariant_utils. Having said that, there are some strange rules about where macros could be and can't be, so it's possible that those rules don't allow a macro to be in zvariant_utils. **If** so, we'll have to live with it being in zvariant_derive (even if it's strange). You should be able to very quickly figure that out when you try to move the macro there.

BTW, the PR title and commit title say "Siganture", while the macro is called "signature".

zeenix commented on zvariant_derive/src/lib.rs:
```diff
@@ -489,3 +490,18 @@ pub fn owned_value_macro_derive(input: TokenStream) -> TokenStream {
+/// Allows to create a `Signature` from a string literal at compile time,
+/// with validation to ensure the signature string is valid.
+///
+/// The macro validates the signature string at compile time:
```
This could just be a copy of the examples in the `Signature` docs itself (after you've converted them to use this macro) + a const example (just copy from the commit log).

zeenix commented:
> Please keep the atomic commits preference in mind, while at it. ;)

:point_up:

hadihaider055 commented:
@zeenix, I've added the documentation with proper examples in the Signature macro and updated the Signature type docs to recommend it. Let me know if there's anything unclear for the readers.

hadihaider055 commented:
> Please keep the atomic commits preference in mind, while at it. ;)

@zeenix, on it! Should I separate it into 3 commits with the feature, docs, and testing? Or rather, the tests be committed with the feature commit `✨ zv: Added a signature macro`.
